### PR TITLE
 Evolve enrollment error handling sdk-294

### PR DESCRIPTION
**Pull request by dmose**
This changes the evolve_enrollments error handling to discard invalid records, rather than simplying blowing up and letting the error propagate up the stack.

It has an integration test for apply_updates.  I thought about writing one for set_global_opt_out, but I'm not convinced it provides much additional information, at least as relevant to this patch.  Which is to say that there shouldn't be errors that such a test would catch that this patch fixes and wouldn't already be caught by the apply_updates integration test. 

### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
  - /No breaking changes/
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.
  - /No new dependencies/

### CODE REVIEW

File: components/nimbus/src/enrollment.rs [hunk 1]
```diff
@@ -2077,6 +2089,63 @@ mod tests {
...
+            enrollments.len(),
+            0,
+            "no new enrollments should have been returned"
+        );
+
+        // XXX is this right?  should we be returning some sort of event for this?
+        assert_eq!(
+            events.len(),
+            0,
+            "no new enrollments should have been returned"
```
Comment (dmose) [inline components/nimbus/src/enrollment.rs:2126]: I'm wondering if maybe we should be returning some sort of event for this.  Thoughts?

File: components/nimbus/tests/common/mod.rs [hunk 2]
```diff
@@ -7,13 +7,25 @@ use nimbus::{error::Result, AppContext, NimbusClient, RemoteSettingsConfig};
...
 #[allow(dead_code)] // not clear why this is necessary...
 pub fn new_test_client(identifier: &str) -> Result<NimbusClient> {
-    use std::path::PathBuf;
     use tempdir::TempDir;
+    let tmp_dir = TempDir::new(identifier)?;
+
+    new_test_client_internal(&tmp_dir)
+}
+
+#[allow(dead_code)] // not clear why this is necessary...
```
Comment (travis79) [inline components/nimbus/tests/common/mod.rs:16]: `[allow(dead_code)]` is needed here as a workaround for https://github.com/rust-lang/rust/issues/46379

Perhaps you could update the `// not clear why this is necessary` comments and redirect to the rust bug/issue?

File: components/nimbus/tests/test_updates.rs [hunk 3]
```diff
@@ -108,4 +111,385 @@ mod test {
...
+                "isEnrollmentPaused":false,
+                "proposedEnrollment":7,
+                "userFacingDescription":"This is a test experiment for diagnostic purposes.",
+            }),
+        ]
+    }
+
+    #[cfg(feature = "rkv-safe-mode")]
+    #[test]
+
```
Comment (travis79) [inline components/nimbus/tests/test_updates.rs:417]: nit: put comment above annotations and remove the blank line

**Review by travis79 — APPROVED** [review]
Just a couple of nits, but this looks good to me!

File: components/nimbus/src/persistence.rs [hunk 4]
```diff
@@ -270,9 +270,11 @@ impl Database {
         match db_version {
             Some(DB_VERSION) => {
                 // Already at the current version, no migration required.
+                log::info!("Already at version {}, no upgrade needed", DB_VERSION);
                 return Ok(());
             }
             Some(1) => {
+                log::info!("Migrating database from v1 to v2");
```
[the remark this answers is not in the document]
Comment (dmose) [reply components/nimbus/src/persistence.rs:277]: Good idea; SDK-308 filed.

File: components/nimbus/src/enrollment.rs (hunk 1 above)
Comment (dmose) [reply components/nimbus/src/enrollment.rs:2125]: SDK-309 filed, and I've mentioned it in the error logging comments.

File: components/nimbus/tests/common/mod.rs (hunk 2 above)
Comment (dmose) [reply components/nimbus/tests/common/mod.rs:16]: Done for all experiments in this file.

File: components/nimbus/tests/test_updates.rs (hunk 3 above)
Comment (dmose) [reply components/nimbus/tests/test_updates.rs:417]: done.

**Review by travis79 — APPROVED** [review]

**Pull request merged**